samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java

```diff
@@ -408,6 +418,43 @@ private void endOfStream() {
     /**
      * Process asynchronously. The callback needs to be fired once the processing is done.
      */
+    private void timer() {
+      state.startTimer();
+      Runnable windowWorker = new Runnable() {
+        @Override
+        public void run() {
+          try {
+            containerMetrics.windows().inc();
+
+            ReadableCoordinator coordinator = new ReadableCoordinator(task.taskName());
+            long startTime = clock.nanoTime();
+            task.window(coordinator);
+            containerMetrics.windowNs().update(clock.nanoTime() - startTime);
+            coordinatorRequests.update(coordinator);
+
+            state.doneWindowOrCommit();
```

Comment:
Is TIMER the same as WINDOW or COMMIT??? Isn't TIMER more like a delayed PROCESS, instead of WINDOW or COMMIT that needs to be exclusive? In fact, the only reason that we made WINDOW mutually exclusive w/ PROCESS is to honor some existing user's read-after-write consistency logic they have assumed in sync RunLoop where process() has to complete before window() is called. However, timer based task should be considered new API and we can remove that guarantee. Maybe we should explicitly separate task.window() and task.onTimer().